### PR TITLE
Reinstated obedient commands for DB

### PR DESCRIPTION
**Pull request by rozza**
JAVA-1784

[Review on Reviewable](https://reviewable.io/reviews/rozza/mongo-java-driver/89)

### CODE REVIEW

File: driver/src/main/com/mongodb/DB.java [hunk 1]
```diff
@@ -617,4 +619,73 @@ private BsonDocument wrap(final DBObject document, final DBEncoder encoder) {
...
+    /**
+     * Determines the read preference that should be used for the given command.
+     *
+     * @param command             the {@link DBObject} representing the command
+     * @param requestedPreference the preference requested by the client.
+     * @return the read preference to use for the given command.  It will never return {@code null}.
+     * @see com.mongodb.ReadPreference
+     */
+    ReadPreference getCommandReadPreference(final DBObject command, final ReadPreference requestedPreference) {
+        if (mongo.getReplicaSetStatus() == null) {
```
Comment (jyemin) [inline driver/src/main/com/mongodb/DB.java]: Is this conditional necessary anymore?  I think command operation will take care of this.

Comment (rozza) [inline driver/src/main/com/mongodb/DB.java]: Excellent I'll remove if not.

File: driver/src/main/com/mongodb/DB.java [hunk 2]
```diff
@@ -617,4 +619,73 @@ private BsonDocument wrap(final DBObject document, final DBEncoder encoder) {
...
+        String comString = command.keySet().iterator().next().toLowerCase();
+
+        if (comString.equals("getnonce") || comString.equals("authenticate")) {
+            return ReadPreference.primaryPreferred();
+        }
+
+        boolean primaryRequired;
+
+        // explicitly check mapreduce commands are inline
+        if (comString.equals("mapreduce")) {
```
Comment (jyemin) [inline driver/src/main/com/mongodb/DB.java]: It's safe to assume that map-reduce and aggregate are done through the helper methods, which already take care of this.  I'd prefer to remove it.

File: driver/src/main/com/mongodb/DB.java [hunk 3]
```diff
@@ -617,4 +619,73 @@ private BsonDocument wrap(final DBObject document, final DBEncoder encoder) {
...
+            if (requestedPreference == null) {
+                return ReadPreference.primary();
+            } else {
+                return requestedPreference;
+            }
+        }
+
+        String comString = command.keySet().iterator().next().toLowerCase();
+
+        if (comString.equals("getnonce") || comString.equals("authenticate")) {
```
Comment (jyemin) [inline driver/src/main/com/mongodb/DB.java]: This was here in 2.x to support DB.authenticate, which no longer exists.  I think we can remove.

File: driver/src/main/com/mongodb/DB.java [hunk 4]
```diff
@@ -617,4 +619,73 @@ private BsonDocument wrap(final DBObject document, final DBEncoder encoder) {
...
+                primaryRequired = outMap.get("inline") == null;
+            } else {
+                primaryRequired = true;
+            }
+        } else if (comString.equals("aggregate")) {
+            @SuppressWarnings("unchecked")
+            List<DBObject> pipeline = (List<DBObject>) command.get("pipeline");
+            primaryRequired = pipeline.get(pipeline.size() - 1).get("$out") != null;
+        } else {
+            primaryRequired = !OBEDIENT_COMMANDS.contains(comString);
```
Comment (jyemin) [inline driver/src/main/com/mongodb/DB.java]: Which just leaves this one-liner

**Comment by rozza** [comment]
Rebased and merged

**Pull request closed without merging**